README
urlclustering
=============

This package facilitates the clustering of similar URLs of a website.

You give a (preferably long and complete) list of URLs as input e.g.:

urls = [
    'http://example.com',
    'http://example.com/about',
    'http://example.com/contact',

    'http://example.com/cat/sports',
    'http://example.com/cat/tech',
    'http://example.com/cat/life',
    'http://example.com/cat/politics',

    'http://example.com/tag/623/tag1',
    'http://example.com/tag/335/tag2',
    'http://example.com/tag/671/tag3',

    'http://example.com/article/?id=1',
    'http://example.com/article/?id=2',
    'http://example.com/article/?id=3',
]

You get a list of clusters as a result. For each cluster you get:
- a REGEX that matches all cluster URLs
- a HUMAN readable string representing the cluster
- a list with all matched cluster URLs

So for our example the result is:

    REGEX: http://example.com/cat/([^/]+)
    HUMAN: http://example.com/cat/[...]
    URLS:
        http://example.com/cat/sports
        http://example.com/cat/tech
        http://example.com/cat/life
        http://example.com/cat/politics

    REGEX: http://example.com/tag/(\d+)/([^/]+)
    HUMAN: http://example.com/tag/[NUMBER]/[...]
    URLS:
        http://example.com/tag/623/tag1
        http://example.com/tag/335/tag2
        http://example.com/tag/671/tag3

    REGEX: http://example.com/article/?\?id=(\d+)
    HUMAN: http://example.com/article?id=[NUMBER]
    URLS:
        http://example.com/article/?id=1
        http://example.com/article/?id=2
        http://example.com/article/?id=3

    UNCLUSTERED URLS:
        http://example.com
        http://example.com/about
        http://example.com/contact


When to use
-----------

This is most useful for website analysis tools that report findings to the user.
E.g. a service that crawls your website and reports page loading time may find
that 10,000 pages take >2 seconds to load. Instead of listing 10,000 URLs it's
better to cluster them. So the end user will see something like:

Slow pages (>2 secs):
- http://example.com/                             (1 URL)
- http://example.com/sitemap                      (1 URL)
- http://example.com/search?q=[...]               (578 URLs)
- http://example.com/tags?tag1=[...]&tag2=[...]   (409 URLs)
- http://example.com/article?id=[NUMBER]          (7209 URLs)


How it works:
-------------
URLs are grouped by domain. Only same domain URLs are clustered.

URLs are then grouped by a signature which is the number of path elements
and the number of QueryString parameters & values the URL has.
Examples:
http://ex.com/about has a signature of (1, 0)
http://ex.com/article?123 has a signature of (1, 1)
http://ex.com/path/to/file?par1=val1&par2=val2 has a signature of (3, 4)

URLs with the same signature are inserted in a tree structure. For
each part (path element or QS parameter or QS value) two nodes are created:
- One with the verbatim part.
- One with the reduced part i.e. a regex that could replace the part.
Leaf nodes hold the number of URLs that match and the number of reductions.

E.g. inserting URL `http://ex.com/article?123` will create 2 top nodes:
    root 1: `article`
    root 2: `[^/]+`
And each top node will have two children:
    child 1: `123`
    child 2: `\d+`
Inserting 3 URLs of the form `/article/[0-9]+` would lead to a tree like this:

           `article`                        `[^/]+`
      /    /      \     \             /    /      \     \
  `123`  `456`  `789`  `\d+`      `123`  `456`  `789`  `\d+`
  1 URL  1 URL  1 URL  3 URLs     1 URL  1 URL  1 URL  3 URLs
  0 re   0 re   0 re   1 re       1 re   1 re   1 re   2  re

The final step is to choose the best leafs. In this case `article` -> `\d+`
is best because it macthes all 3 URLs with 1 reduction so the cluster returned
is http://ex.com/article?(\d+)


License
-------

Copyright (c) 2015 Dimitris Giannitsaros.

Licensed under the MIT License.

CHANGES
urlclustering changelog
=======================


Version 0.4.1
-------------
Removed unicode from setup.py


Version 0.4
-----------
Python 3 support.


Version 0.3
-----------
Using rst.


Version 0.2
-----------
Docs updated.


Version 0.1
-----------
First public release.

urlclustering/tests/test_urlclustering.py
#!/usr/bin/env python
from __future__ import unicode_literals
from urlclustering.urlclusterer import cluster_urls
from urlclustering.reimprover import improve_patterns, _str_reduce
import unittest


class TestUrlClusterer(unittest.TestCase):

    def test_single_cluster(self):
        urls = ['http://s.com/blah/%d' % x for x in range(1, 20)]
        c = cluster_urls(urls, 10)
        self.assertEqual(c['unclustered'], [])
        self.assertEqual(
            sorted(c['clusters'].keys()),
            sorted([('http://s.com/blah/(\\d+)',
                     'http://s.com/blah/[NUMBER]')]))
        self.assertEqual(
            sorted(c['clusters'].values()), sorted([urls]))

    def test_mixed(self):
        c_urls = ['http://s.com/blah/?id=%d' % x for x in range(1, 20)]
        u_urls = ['http://s.com/asdf', 'http://s.com/a/a/b']
        c = cluster_urls(c_urls + u_urls, 10)
        self.assertEqual(sorted(c['unclustered']), sorted(u_urls))
        self.assertEqual(
            sorted(c['clusters'].keys()),
            sorted([('http://s.com/blah/?\\?id=(\\d+)',
                     'http://s.com/blah?id=[NUMBER]')]))
        self.assertEqual(
            sorted(c['clusters'].values()), sorted([c_urls]))

    def test_other(self):
        x_urls = ['http://s.com/blah/%d' % x for x in range(1, 20)]
        y_urls = ['http://s.com/a/b/aa%dbb' % x for x in range(1, 20)]
        z_urls = ['http://b.com/ab/aa%dbb' % x for x in range(1, 50)]
        c = cluster_urls(x_urls + y_urls + z_urls, 10)
        improve_patterns(c['clusters'])
        self.assertEqual(c['unclustered'], [])
        self.assertEqual(
            sorted(c['clusters'].keys()),
            sorted([('http://b.com/ab/aa([^/]+)bb',
                     'http://b.com/ab/aa[...]bb'),
                    ('http://s.com/blah/(\\d+)',
                     'http://s.com/blah/[NUMBER]'),
                    ('http://s.com/a/b/aa([^/]+)bb',
                     'http://s.com/a/b/aa[...]bb')]))

    def test_lazy_str_reduce(self):
        return
        strings = ['abc123asdf', 'abc1234asdf', 'abc234asdf']
        self.assertEqual(_str_reduce(strings, '(.*)', '[...]'),
                         ('abc(.*)asdf', 'abc[...]asdf'))

    def test_lazy_str_reduce_other(self):
        return
        strings = ['1abc123asdf3', '2abc1234asdf4', '3abc234asdf5']
        self.assertEqual(_str_reduce(strings, '(.*)', '[...]'),
                         ('(.*)', '[...]'))


urlclustering/urltree.py
from __future__ import unicode_literals
import re


class URLTreeNode(dict):
    """
    Tree structure based on dictionary. Leaf nodes define a regular
    expression that matches certain URLs (also given in human readable form).
    These regural expressions vary from strict e.g. ~/a/b/1~ (that match 1 URL)
    to the very generic e.g. ~[^/]+/[^/]+/\d+~ (that match all URLs).
    """

    REDUCED_NUM_LITERAL = '(\d+)'
    REDUCED_PATH_LITERAL = '([^/]+)'
    REDUCED_PARAM_LITERAL = '([^&=?]+)'

    def __init__(self, iterable=(), r='', h='', reductions=0):
        self.r = r  # regular expression so far
        self.h = h  # human readable expression so far
        self.urls = set()  # set of urls matched by self.r (leafs only)
        self.reductions = reductions  # reductions to root (leafs only)
        dict.__init__(self, iterable)

    def __repr__(self):
        if len(self) == 0:
            return '%s -> %s urls with %s reductions' % (
                self.r, len(self.urls), self.reductions)
        return '\n'.join(repr(self[k]) for k in self)

    def leafs(self):
        leafs = []
        if len(self) == 0:
            # leaf
            return [{'pattern': self.r,
                     'h_pattern': self.h,
                     'urls': self.urls,
                     'reductions': self.reductions}]
        for key in self:
            leafs.extend(self[key].leafs())
        return leafs

    def _reduced(self, elem, separator):
        """Return regex and human readable expressions"""
        if re.search(r'^\d+$', elem):
            return URLTreeNode.REDUCED_NUM_LITERAL, '[NUMBER]'
        elif separator == '/':
            return URLTreeNode.REDUCED_PATH_LITERAL, '[...]'
        else:
            return URLTreeNode.REDUCED_PARAM_LITERAL, '[...]'

    def _re_sep(self, separator):
        return '/?\?' if separator == '?' else separator

    def add_url(self, parsed_url, position=0):
        if position >= len(parsed_url.parts):
            # leaf node
            self.urls.add(parsed_url.url)
        else:
            sep, elem = parsed_url.parts[position]
            # add part as is
            if elem not in self:
                self[elem] = URLTreeNode(
                    r='%s%s%s' % (self.r, self._re_sep(sep), re.escape(elem)),
                    h='%s%s%s' % (self.h, sep, elem),
                    reductions=self.reductions
                )
            self[elem].add_url(parsed_url, position+1)
            # add reduced part
            reduced, reduced_h = self._reduced(elem, sep)
            if reduced not in self:
                self[reduced] = URLTreeNode(
                    r='%s%s%s' % (self.r, self._re_sep(sep), reduced),
                    h='%s%s%s' % (self.h, sep, reduced_h),
                    reductions=self.reductions+1
                )
            self[reduced].add_url(parsed_url, position+1)


urlclustering/parsedurl.py
from __future__ import unicode_literals
import re


class ParsedURL():
    """
    Parses a URL and exposes:
        - parts: a list with path and QS elements:
            http://url.com/abc/efg?aaa=123&bbb=456 would give
            [('/', 'abc'), ('/', 'efg'), ('?', 'aaa'), ('=', '123'),
             ('&', 'bbb'), ('=', '456')]
        - signature: based on number of path and QS elements. Only same
          signature URLs are clustered
        - domain: actually including everything (protocol, port etc) up to path
    """
    _parts = []
    _signature = (0, 0)  # (path_parts, qs_parts)
    _domain = None

    def __init__(self, url):
        self.url = url
        self._parts = []
        # path parts
        path_re = re.search(r'https?://[^/?#]+/([^?#]+)', url)
        if path_re:
            elems = path_re.group(1).strip('/').split('/')
            self._parts = list(zip(['/'] * len(elems), elems))
        path_parts = len(self._parts)
        # QS parts
        qs_re = re.search(r'https?://[^?#]+\?([^#]+)', url)
        if qs_re:
            elems = qs_re.group(1).strip('=?&').split('&')
            for i, part in enumerate(elems):
                sep = '?' if i == 0 else '&'
                if '=' in part:
                    par, val = part.split('=')
                    self._parts.append((sep, par))
                    self._parts.append(('=', val))
                else:
                    self._parts.append((sep, part))
        # signature
        self._signature = (path_parts, len(self._parts) - path_parts)
        # domain
        self._domain = None
        domain_re = re.search(r'^((https?://)[^/\?\#]+)', url, re.U)
        if domain_re:
            self._domain = domain_re.group(1)

    @property
    def parts(self):
        return self._parts

    @property
    def signature(self):
        return self._signature

    @property
    def domain(self):
        return self._domain


urlclustering/reimprover.py
from __future__ import unicode_literals
import re
from urlclustering.urltree import URLTreeNode

"""
Work in progress - enter at your own risk :)

The idea is to improve a regex for a specific set of urls
e.g. for these URLs
    /path/to/thumb_dog.jpg
    /path/to/thumb_cat.jpg
    /path/to/thumb_hen.jpg
we currently get
    /path/to/([^/]+)
but we can also get
    /path/to/thumb_([^/]+).jpg
"""


def _str_reduce(strings, pattern, h_pattern):
    """
    Reduces the given strings to a regexp by extracting the
    commong prefix and suffix
    """
    prefix = ''
    for ix in range(1, len(strings[0])):
        prefix = strings[0][:ix]
        matching = len(strings) == len([x for x in strings if x[:ix] == prefix])
        if not matching:
            prefix = strings[0][:ix-1]
            break

    suffix = ''
    for ix in range(1, len(strings[0]) - len(prefix)):
        suffix = strings[0][-ix:]
        matching = len(strings) == len([x for x in strings if x[-ix:] == suffix])
        if not matching:
            if ix > 1:
                suffix = strings[0][-(ix-1):]
            else:
                suffix = ''
            break

    return (u"%s%s%s" % (prefix, pattern, suffix),
            u"%s%s%s" % (prefix, h_pattern, suffix))


def _improve_cluster_pattern(cluster_key, urls):
    """
    Improve regex to use maximum literals and minimum metacharacters.
    All urls must be matched.
    """
    pattern, h_pattern = cluster_key
    pattern_parsed = []
    h_pattern_parsed = []
    cur_pattern = pattern
    cur_h_pattern = h_pattern
    while True:
        if cur_pattern == '':
            break
        p = re.search(r'^(.*?)((%s)|(%s))(.*)$' % (
            re.escape(URLTreeNode.REDUCED_PATH_LITERAL),
            re.escape(URLTreeNode.REDUCED_PARAM_LITERAL)), cur_pattern)
        h_p = re.search(r'^(.*?)(\[\.\.\.\])(.*)$', cur_h_pattern)
        if not p:
            if cur_pattern != '':
                pattern_parsed.append(cur_pattern)
                h_pattern_parsed.append(cur_h_pattern)
            break
        if p.group(1) != '':
            pattern_parsed.append(p.group(1))
            h_pattern_parsed.append(h_p.group(1))
        pattern_parsed.append(p.group(2))
        h_pattern_parsed.append(h_p.group(2))
        cur_pattern = p.group(5)
        cur_h_pattern = h_p.group(3)

    group = 0
    matches = [re.search(pattern, url) for url in urls]
    improved_pattern = ''
    improved_h_pattern = ''
    for ix in range(0, len(pattern_parsed)):
        if (pattern_parsed[ix] !=
            URLTreeNode.REDUCED_PATH_LITERAL and
            pattern_parsed[ix] !=
                URLTreeNode.REDUCED_PARAM_LITERAL):
            improved_pattern += pattern_parsed[ix]
            improved_h_pattern += h_pattern_parsed[ix]
            continue
        group += 1
        reduced_strings = [m.group(group) for m in matches]
        improved = _str_reduce(
            reduced_strings, pattern_parsed[ix], h_pattern_parsed[ix])
        improved_pattern += improved[0]
        improved_h_pattern += improved[1]

    return (improved_pattern, improved_h_pattern)


def improve_patterns(clusters):
    for key in clusters.keys():
        improved = _improve_cluster_pattern(key, clusters[key])
        if improved != key:
            clusters[improved] = clusters[key]
            del clusters[key]


urlclustering/urlclusterer.py
from __future__ import unicode_literals
import re
from collections import defaultdict
from urlclustering.urltree import URLTreeNode
from urlclustering.parsedurl import ParsedURL

"""
This package facilitates the clustering of similar URLs.

How it works:
-------------
URLs are grouped by domain. Only same domain URLs are clustered.

URLs are then grouped by a signature which is the number of path elements
and the number of QueryString parameters & values the URL has.
Examples:
http://ex.com/about has a signature of (1, 0)
http://ex.com/article?123 has a signature of (1, 1)
http://ex.com/path/to/file?par1=val1&par2=val2 has a signature of (3, 4)

URLs with the same signature are inserted in a tree structure. For
each part (path element or QS parameter or QS value) two nodes are created:
- One with the verbatim part.
- One with the reduced part i.e. a regex that could replace the part.
Leaf nodes hold the number of URLs that match and the number of reductions.

E.g. inserting URL `http://ex.com/article?123` will create 2 top nodes:
    root 1: `article`
    root 2: `[^/]+`
And each top node will have two children:
    child 1: `123`
    child 2: `\d+`
Inserting 3 URLs of the form `/article/[0-9]+` would lead to a tree like this:

           `article`                        `[^/]+`
      /    /      \     \             /    /      \     \
  `123`  `456`  `789`  `\d+`      `123`  `456`  `789`  `\d+`
  1 URL  1 URL  1 URL  3 URLs     1 URL  1 URL  1 URL  3 URLs
  0 re   0 re   0 re   1 re       1 re   1 re   1 re   2  re

The final step is to choose the best leafs. In this case `article` -> `\d+`
is best because it macthes all 3 URLs with 1 reduction so the cluster returned
is http://ex.com/article?(\d+)
"""


def _cluster_same_signature_urls(parsed_urls, min_cluster_size):
    """
    Returns patterns matching >=min_cluster_size urls in best -> worst order.
    List of URLs must have the same signature (path + QS elements).
    """
    patterns = []
    if len(parsed_urls) == 0:
        return patterns
    max_reductions = len(parsed_urls[0].parts)

    # build our URL tree
    root = URLTreeNode()
    for parsed in parsed_urls:
        root.add_url(parsed)

    # reduce leafs by removing the best one at each iteration
    leafs = root.leafs()
    while leafs:
        bestleaf = max(
            leafs,
            key=lambda x:
                len(x['urls']) * (max_reductions - x['reductions']) ** 2
        )
        if len(bestleaf['urls']) >= min_cluster_size:
            patterns.append((bestleaf['pattern'],
                             bestleaf['h_pattern']))
        leafs.remove(bestleaf)
        remaining_leafs = []
        for leaf in leafs:
            leaf['urls'] -= bestleaf['urls']
            if leaf['urls']:
                remaining_leafs.append(leaf)
        leafs = remaining_leafs

    return patterns


def _cluster_same_domain_urls(parsed_urls, min_cluster_size):
    """Returns clusters and a list of unclustered urls."""

    # group URLs by signature
    url_map = defaultdict(list)
    for parsed in parsed_urls:
        url_map[parsed.signature].append(parsed)

    # build clusters
    clusters = defaultdict(list)
    unclustered = []
    for parsed_urls in url_map.values():
        if len(parsed_urls) < min_cluster_size:
            unclustered.extend([x.url for x in parsed_urls])
            continue
        patterns = _cluster_same_signature_urls(parsed_urls, min_cluster_size)
        for (pattern, h_pattern) in patterns:
            remaining_urls = []
            # add matching URLs to cluster and remove from remaining URLs
            for parsed in parsed_urls:
                if re.search(pattern, parsed.url):
                    clusters[(pattern, h_pattern)].append(parsed.url)
                else:
                    remaining_urls.append(parsed)
            parsed_urls = remaining_urls

        # everything left goes to unclustered
        unclustered.extend(x.url for x in parsed_urls)

    return {'clusters': clusters, 'unclustered': unclustered}


def cluster_urls(urls, min_cluster_size=10):
    if min_cluster_size < 2:
        min_cluster_size = 2
    res = {'clusters': {}, 'unclustered': []}

    # get ParsedURL objects for each url
    parsed_urls = []
    for url in urls:
        parsed_urls.append(ParsedURL(url))

    # group urls by domain
    by_domain = defaultdict(list)
    for parsed in parsed_urls:
        try:
            by_domain[parsed.domain].append(parsed)
        except:
            res['unclustered'].append(parsed.url)

    # cluster in each domain group
    for domain, parsed_urls in by_domain.items():
        c_res = _cluster_same_domain_urls(parsed_urls, min_cluster_size)
        res['clusters'].update({('%s%s' % (domain, k[0]),
                                 '%s%s' % (domain, k[1])): v
                                for k, v in c_res['clusters'].items()})
        res['unclustered'].extend(c_res['unclustered'])

    return res
